# Tomato-root-atlas-2020
Code and resources for Tomato Root Atlas (2020)

**Title:** Novelty, conservation, and repurposing of gene function in root cell type development

**Authors:**  Kaisa Kajala, Mona Gouran, Lidor Shaar-Moshe, G. Alex Mason, Joel Rodriguez-Medina1, Dorota Kawa, Germain Pauluzzi, Mauricio Reynoso, Alex Canto-Pastor, Concepcion Manzano, Vincent Lau, Mariana A. S. Artur, Donnelly A. West, Sharon B. Gray, Alexander T. Borowsky, Bryshal P. Moore, Andrew I. Yao, Kevin W. Morimoto, Marko Bajic, Elide Formentin, Niba Nirmal, Alan Rodriguez, Asher Pasha, Roger B. Deal, Daniel Kliebenstein, Torgeir R. Hvidsten, Nicholas J. Provart, Neelima Sinha, Daniel E. Runcie, Julia Bailey-Serres, Siobhan M. Brady

**Journal:** TBC

**Abstract:** 

> TBC


## Code used by topic (*in progress*)
- [Obtaining RNA-seq read counts: Trimming, Kallisto and STAR](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Scripts/Trimming_Kallist_STAR_mapping.sh)
- [Relative differential expression: Roku](https://github.com/plant-plasticity/tomato-root-atlas-2020/tree/master/Scripts/Roku)
- [Co-expression network analysis: WGCNA ](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Scripts/WGCNA-Tomato-ATLAS.rmd)
- [More to come...]

## Protocols
These detailed bench protocols describe the methods as carried out in the Tomato Root Atlas paper.
- [TRAP purification of translating ribosomes](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Protocols/TRAP%20Atlas.pdf)
- [RNA-seq library preparation from TRAP](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Protocols/Brads%20Rapid%20Ravi%20for%20low%20mRNA%20Atlas.pdf)
- [INTACT isolation of nuclei](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Protocols/INTACT%20Atlas.pdf)
- [ATAC-seq library preparation from INTACT](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Protocols/ATAC-seq%20library%20prep%20Atlas.pdf)

## Supplementary data
- [Pairwise comparison of ATAC cut-counts between cell types](https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Figures/Figure_S23_scatter_plot_replicates_repUnion_THSs_ALL_110520_v2_with_legend.pdf)

## Other resources by topic (*in progress*)
- [Data on GEO]
- [eFP browser]


## Analysis pipeline summaries
- **Flowchart for multispecies analysis.** An overview of the pipeline and methods used for exploring the conservation of homologous cell types and tissues among tomato, Arabidopsis and rice. Ath=Arabidopsis thaliana; Osa=Oryza sativa; Sly=Solanum lycopersicum; IQR= interquartile range; TPM=transcript per million; GO=gene ontology; CTEG=cell type/tissue enriched genes. 

   <img src="https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Figures/Figure_S13_multispecies%20overview.jpg" width="700">


- **Flowchart for identification and annotation of ATAC-seq data.** Data analysis overview for methods used for identification of replicate transposase hypersensitive sites. 

   <img src="https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Figures/Figure_S20_ATACseq_flowchart.jpg" width="700">


- **Unique cell type network construction pipeline.** Data analysis overview for methods used to construct inferred cell type-unique regulatory networks. 

   <img src="https://github.com/plant-plasticity/tomato-root-atlas-2020/blob/master/Figures/Figure_S26_NetworkFlowchart.jpg" width="700">

